src/OrleansServiceFabricUtils/ServiceFabricNamingServiceGatewayProvider.cs

```diff
@@ -0,0 +1,372 @@
+
+        private const string ETagPrefix = ".ETAG_";
+
+        private readonly FabricClient fabricClient;
+
+        private Logger log;
+
+        private FabricClient.PropertyManagementClient store;
+
+        private Uri tableUri;
+
+        public ServiceFabricNamingServiceGatewayProvider(FabricClient client)
+        {
+            this.fabricClient = client;
+        }
+
+        public Task InitializeGatewayListProvider(ClientConfiguration config, Logger logger)
+        {
+            this.Initialize(logger, config.DeploymentId);
+            this.MaxStaleness = config.GatewayListRefreshPeriod;
+            
+            return Task.FromResult(0);
+        }
+
+        public async Task InitializeMembershipTable(GlobalConfiguration config, bool tryInitTableVersion, Logger logger)
+        {
+            this.Initialize(logger, config.DeploymentId);
+            this.log = logger;
+
+            if (tryInitTableVersion)
+            {
+                try
+                {
+                    await this.store.CreateNameAsync(this.tableUri);
+                }
+                catch (FabricElementAlreadyExistsException)
+                {
+                    this.log?.Verbose($"Membership table already exists in property store at {this.tableUri}");
+                }
+
+                var ops = new PropertyBatchOperation[]
+                              {
+                                  // Check preconditions.
+                                  new CheckExistsPropertyOperation(ETagPropertyName, false),
+
+                                  // Update version and insert rows.
+                                  new PutPropertyOperation(VersionPropertyName, 0),
+                                  new PutPropertyOperation(ETagPropertyName, DefaultETag)
+                              };
+                await this.store.SubmitPropertyBatchAsync(this.tableUri, ops);
+            }
+        }
+
+        private void Initialize(Logger logger, string deploymentId)
+        {
+            this.tableUri = GetTableUri(deploymentId);
+            this.log = logger;
+            this.store = this.fabricClient.PropertyManager;
+        }
+
+        private static Uri GetTableUri(string deploymentId)
+        {
+            return new Uri("fabric:/silos_" + deploymentId);
+        }
+
+        public Task<MembershipTableData> ReadRow(SiloAddress siloAddress)
+        {
+            return this.ReadEntries(siloAddress);
+        }
+
+        public Task<MembershipTableData> ReadAll()
+        {
+            return this.ReadEntries();
+        }
+
+        private async Task<MembershipTableData> ReadEntries(SiloAddress siloAddress = null)
+        {
+            ReadResults result;
+            do
+            {
+                // Continue attempting to read from the table until a consistent read is made.
+                result = await this.TryReadEntries(siloAddress);
+            }
+            while (!result.IsConsistent);
```

Comment:
I am sure you know that already, but you could have also integrated differently: instead of using NS as a key value store for MBR table, use NS as a full blown membership provider: listen to its notification about silo liveness, read all silos from it, use the SF internal membership basically and fully replace Orleans' MBR. Deeper integration sort of.
That's what I did 3-4 years ago.